### ISSUE TITLE
Uncaught Invariant Violation: Cannot call hover while not dragging. error in Chrome 77.0 when using iframes

### ISSUE THREAD

**Issue by nemesv** [react-dnd/react-dnd#1534]
Update 2019-10-14
---------------------
**The [bug was in Chrome](https://bugs.chromium.org/p/chromium/issues/detail?id=1005747) and fixed in Chrome version: 77.0.3865.120**
-----

We are using react-dnd in an web application which is hosted inside another webapplication using a cross-origin iframe. 

We have noticed that since the last Chrome upgrade all the drag-drop functionality stopped working. Our app is working fine in older Chrome versions and in other browsers.

We are not sure whether this is hitting an edge case in this library or the issue is with in Chrome 77.0

**Reproduction**

Go to http://react-dnd.github.io/react-dnd/examples/dustbin/single-target (or to any of the examples)

The "inline" example is working fine:
![chrome-working](https://user-images.githubusercontent.com/251330/64860544-c6e0bb80-d62d-11e9-8f0d-fed1c1f7c397.gif)

However running the same code in the JavaScript/TypeScript tabs, so in a cross-origin iframe (e.g. in the embedded codesandbox) the drop event is not fired:

![chrome-not-working](https://user-images.githubusercontent.com/251330/64860595-e841a780-d62d-11e9-9b14-1e421b4c40c1.gif)

[Running the example](https://codesandbox.io/s/github/react-dnd/react-dnd/tree/gh-pages/examples_hooks_ts/01-dustbin/single-target?from-embed) is also not working in the CodeSandBox editor which also uses an inframe.

![chrome-not-working-codesandbox](https://user-images.githubusercontent.com/251330/64860858-bbda5b00-d62e-11e9-833b-664c93989a23.gif)

In both cases there is this error in the console:

```
browser.js:38 Uncaught Invariant Violation: Cannot call hover while not dragging.
    at invariant (https://hmq56.csb.app/node_modules/invariant/browser.js:38:15)
    at checkInvariants (https://hmq56.csb.app/node_modules/dnd-core/dist/esm/actions/dragDrop/hover.js:45:29)
    at DragDropManagerImpl.hover (https://hmq56.csb.app/node_modules/dnd-core/dist/esm/actions/dragDrop/hover.js:25:5)
    at Object.eval [as hover] (https://hmq56.csb.app/node_modules/dnd-core/dist/esm/DragDropManagerImpl.js:132:38)
    at HTML5Backend.handleTopDrop (https://hmq56.csb.app/node_modules/react-dnd-html5-backend/dist/esm/HTML5Backend.js:418:21)
```

**Expected behavior**
No exceptions in the console, drop event is fired

**Desktop (please complete the following information):**

- OS: Windows 10 (1903)
- Browser Chrome
- Version 77.0.3865.75 (Official Build) (64-bit)

**Additional context**
We did some investigation and we've observed that Chrome fires the dragEnd event prematurely in the case of the iframe which breaks the invariant checks.

**Comment by nolans4** [comment]
Seeing the same issues in our app, we also have our DND code within an iframe and its completely broken for users with the newest chrome update.

Errors:
<img width="452" alt="Screen Shot 2019-09-13 at 3 42 54 PM" src="https://user-images.githubusercontent.com/3512114/64871470-39a56300-d63d-11e9-943f-0aa2571d0e1a.png">

**Comment by timmfin** [comment]
This is also affecting some of the drag and drop functionality in the various editor apps at HubSpot (that use react-dnd inside an iframe). Still digging into the extent of the problem.

**Comment by timmfin** [comment]
I haven't had a chance to compare it with the latest HTML5 backend on react-dnd master, but we have deployed a temporary, hacky fix to our custom HTML5 backend:

![image](https://user-images.githubusercontent.com/60455/64888073-08766400-d638-11e9-836f-010833249e08.png)

Anecdotally, that `setTimeout` delay of the `handleTopDragEndCapture` event works 95%+ of the time. Because we have seen an occasional failure where the `Invariant Violation` error still happens, likely because the delay wasn't long enough. However, this at least has improved the situation for us, and we can look at a more robust hack if/when needed.

**Comment by adeelraza** [comment]
+1

**Comment by TheConfuZzledDude** [comment]
I get this same error in Vivaldi 2.8.1664.30 (Official Build) snapshot (64-bit) on every component, not just those in an iframe, whereas Chrome seems to work perfectly. I tried making the change that @timmfin suggested but it did nothing for me

**Comment by jkells** [comment]
This error is pretty devastating for us. We have multiple apps that effectively broke completely over a few days as users updated Chrome.

@timmfin 's workaround is working for us, thanks!

**Comment by jkells** [comment]
I've been playing around with this for a few hours tonight on the master branch. It's not throwing anymore but you're unable to drop any items. The setTimeout fix seems to fix it.

It looks to me like we get a handleTopDragEndCapture followed by a handleDrop then a handleTopDrop

If we call endDrag() within handleTopDragEndCapture the drop events don't fire at all in an IFRAME.

**Comment by jkells** [comment]
Interestingly when not in an IFRAME I only see the dragend events when a drag is cancelled not right before a successful drop. That seems to be the difference.

**Comment by jkells** [comment]
Here we go, I think this is the issue: https://bugs.chromium.org/p/chromium/issues/detail?id=1005747

**Comment by DonMcNamara** [comment]
Has anyone back ported the workaround to older versions of this library? Like 2.6.0 :sweat_smile:

**Comment by TheConfuZzledDude** [comment]
As a workaround, just for desktop use, I've been using https://github.com/zyzo/react-dnd-mouse-backend as an alternative to the HTML5 backend. It seems to work fine for me

**Comment by DonMcNamara** [comment]
Just FYI, I'm seeing this on Linux now as well.

**Comment by vvitto** [comment]
@timmfin Good job, this works for me!) I've prepared a temporary package with this fix. So anyone can just replace the old one for a while. https://www.npmjs.com/package/@vvitto/html5-backend-chrome-dnd-fix

**Comment by Cristy94** [comment]
> As a workaround, just for desktop use, I've been using https://github.com/zyzo/react-dnd-mouse-backend as an alternative to the HTML5 backend. It seems to work fine for me

I also tried this fix, but the mouse backend is really buggy in FireFox.

**Comment by TheConfuZzledDude** [comment]
@Cristy94 It seems that he deleted the comment, but https://www.npmjs.com/package/@vvitto/html5-backend-chrome-dnd-fix was posted before. I haven't tested it, but it might work?

**Comment by vvitto** [comment]
@TheConfuZzledDude @Cristy94 it works for me, but you should check the version of your backend. It works fine on 2.5.4

**Comment by Cristy94** [comment]
> @TheConfuZzledDude @Cristy94 it works for me, but you should check the version of your backend. It works fine on 2.5.4

Does the example mentioned in the first post work for you when inside the editor? It does not work for on Windows 10 - Chrome 77

**Comment by Cristy94** [comment]
> @TheConfuZzledDude @Cristy94 it works for me, but you should check the version of your backend. It works fine on 2.5.4

What works? Mouse backend doesn't have version 2.5.4 https://github.com/zyzo/react-dnd-mouse-backend/releases and the html5 backend doesn't work on Windows 10 - Chrome 77 as mentioned in the original post.

**Comment by acroyear** [comment]
> Has anyone back ported the workaround to older versions of this library? Like 2.6.0 😅

yes, we did some adaptation of the workaround to html5-backend 2.2.7 and it worked well enough to put into production.

**Comment by acroyear** [comment]
> Just FYI, I'm seeing this on Linux now as well.

Chromium's bug acknowledged that Linux has the same, and it will be fixed with the Windows version.  Latest state is M78 has it in the beta coming out in a few days, M79 and Canary have it already.  They're still discussing patching M77.

**Comment by purpletastes** [comment]
> 我还没有机会将它与react-dnd master上的最新HTML5后端进行比较，但是我们已经对自定义HTML5后端部署了一个临时的，hacky的修复程序：
> 
> ![图片](https://user-images.githubusercontent.com/60455/64888073-08766400-d638-11e9-836f-010833249e08.png)
> 
> 有趣的是，事件的`setTimeout`延迟在`handleTopDragEndCapture`95％以上的时间内有效。因为我们看到偶尔的故障`Invariant Violation`仍然会发生错误，可能是因为延迟时间不够长。但是，这至少为我们改善了情况，如果需要的话，我们可以考虑采用更强大的技术。

这个问题现在仍然存在，我怎么样使用才能够解决呢，急急急！

**Comment by purpletastes** [comment]
> 我们在一个Web应用程序中使用react-dnd，该Web应用程序使用跨源iframe托管在另一个Web应用程序中。
> 
> 我们已经注意到，自从上次Chrome升级以来，所有拖放功能均停止工作。我们的应用程序在较旧的Chrome版本和其他浏览器中均可正常运行。
> 
> 我们不确定这是否在该库中处于领先地位，还是在Chrome 77.0中出现问题
> 
> **再生产**
> 
> 转到http://react-dnd.github.io/react-dnd/examples/dustbin/single-target（或任何示例）
> 
> 重现行为的步骤：
> 
> “内联”示例运行良好：
> ![镀铬](https://user-images.githubusercontent.com/251330/64860544-c6e0bb80-d62d-11e9-8f0d-fed1c1f7c397.gif)
> 
> 但是，在JavaScript / TypeScript标签中运行相同的代码，因此在跨域iframe中（例如，在嵌入式codeandbox中），不会触发drop事件：
> 
> ![铬不工作](https://user-images.githubusercontent.com/251330/64860595-e841a780-d62d-11e9-9b14-1e421b4c40c1.gif)
> 
> [运行示例](https://codesandbox.io/s/github/react-dnd/react-dnd/tree/gh-pages/examples_hooks_ts/01-dustbin/single-target?from-embed)也无法在CodeSandBox编辑器中使用，后者也使用内置框架。
> 
> ![chrome-not-working-codesandbox](https://user-images.githubusercontent.com/251330/64860858-bbda5b00-d62e-11e9-833b-664c93989a23.gif)
> 
> 在这两种情况下，控制台中都会出现此错误：
> 
> ```
> browser.js:38 Uncaught Invariant Violation: Cannot call hover while not dragging.
>     at invariant (https://hmq56.csb.app/node_modules/invariant/browser.js:38:15)
>     at checkInvariants (https://hmq56.csb.app/node_modules/dnd-core/dist/esm/actions/dragDrop/hover.js:45:29)
>     at DragDropManagerImpl.hover (https://hmq56.csb.app/node_modules/dnd-core/dist/esm/actions/dragDrop/hover.js:25:5)
>     at Object.eval [as hover] (https://hmq56.csb.app/node_modules/dnd-core/dist/esm/DragDropManagerImpl.js:132:38)
>     at HTML5Backend.handleTopDrop (https://hmq56.csb.app/node_modules/react-dnd-html5-backend/dist/esm/HTML5Backend.js:418:21)
> ```
> 
> **预期行为**
> 控制台中无异常，引发了drop事件
> 
> **桌面（请完成以下信息）：**
> 
> * 作业系统：Windows 10（1903）
> * 浏览器Chrome
> * 版本77.0.3865.75（官方内部版本）（64位）
> 
> **附加上下文**
> 我们进行了一些调查，我们发现在iframe打破不变检查的情况下，Chrome会过早触发dragEnd事件。

我也希望能够尽快解决这个问题，chrome版本77出现这个问题

**Comment by purpletastes** [comment]
> 这也影响到HubSpot的各种编辑器应用程序（在iframe中使用react-dnd）的某些拖放功能。仍在深入研究问题的程度。
> 
> 我已经复制了Windows 10的Chrome 77.0.3865.75中的[codeandbox示例所](https://codesandbox.io/s/github/react-dnd/react-dnd/tree/gh-pages/examples_hooks_ts/01-dustbin/single-target?from-embed)演示的问题。

怎么解决的，麻烦回复一下，谢谢

**Comment by purpletastes** [comment]
> This is also affecting some of the drag and drop functionality in the various editor apps at HubSpot (that use react-dnd inside an iframe). Still digging into the extent of the problem.
> 
> And I have replicated the problem demonstrated by the [codesandbox example](https://codesandbox.io/s/github/react-dnd/react-dnd/tree/gh-pages/examples_hooks_ts/01-dustbin/single-target?from-embed) in Chrome 77.0.3865.75 on Windows

> We are using react-dnd in an web application which is hosted inside another webapplication using a cross-origin iframe.
> 
> We have noticed that since the last Chrome upgrade all the drag-drop functionality stopped working. Our app is working fine in older Chrome versions and in other browsers.
> 
> We are not sure whether this is hitting an edge case in this library or the issue is with in Chrome 77.0
> 
> **Reproduction**
> 
> Go to http://react-dnd.github.io/react-dnd/examples/dustbin/single-target (or to any of the examples)
> 
> Steps to reproduce the behavior:
> 
> The "inline" example is working fine:
> ![chrome-working](https://user-images.githubusercontent.com/251330/64860544-c6e0bb80-d62d-11e9-8f0d-fed1c1f7c397.gif)
> 
> However running the same code in the JavaScript/TypeScript tabs, so in a cross-origin iframe (e.g. in the embedded codesandbox) the drop event is not fired:
> 
> ![chrome-not-working](https://user-images.githubusercontent.com/251330/64860595-e841a780-d62d-11e9-9b14-1e421b4c40c1.gif)
> 
> [Running the example](https://codesandbox.io/s/github/react-dnd/react-dnd/tree/gh-pages/examples_hooks_ts/01-dustbin/single-target?from-embed) is also not working in the CodeSandBox editor which also uses an inframe.
> 
> ![chrome-not-working-codesandbox](https://user-images.githubusercontent.com/251330/64860858-bbda5b00-d62e-11e9-833b-664c93989a23.gif)
> 
> In both cases there is this error in the console:
> 
> ```
> browser.js:38 Uncaught Invariant Violation: Cannot call hover while not dragging.
>     at invariant (https://hmq56.csb.app/node_modules/invariant/browser.js:38:15)
>     at checkInvariants (https://hmq56.csb.app/node_modules/dnd-core/dist/esm/actions/dragDrop/hover.js:45:29)
>     at DragDropManagerImpl.hover (https://hmq56.csb.app/node_modules/dnd-core/dist/esm/actions/dragDrop/hover.js:25:5)
>     at Object.eval [as hover] (https://hmq56.csb.app/node_modules/dnd-core/dist/esm/DragDropManagerImpl.js:132:38)
>     at HTML5Backend.handleTopDrop (https://hmq56.csb.app/node_modules/react-dnd-html5-backend/dist/esm/HTML5Backend.js:418:21)
> ```
> 
> **Expected behavior**
> No exceptions in the console, drop event is fired
> 
> **Desktop (please complete the following information):**
> 
> * OS: Windows 10 (1903)
> * Browser Chrome
> * Version 77.0.3865.75 (Official Build) (64-bit)
> 
> **Additional context**
> We did some investigation and we've observed that Chrome fires the dragEnd event prematurely in the case of the iframe which breaks the invariant checks.

请问这个问题解决了吗？

**Comment by acroyear** [comment]
The original Chrome bug has been merged to the upcoming M77 patch release. I can not say when it will actually come out, just that I've been following the chromium bug page on this closely.

**Comment by nemesv** [comment]
The bug was in Chrome and fixed in Chrome version: 77.0.3865.120

**Comment by Kadrian** [comment]
Chrome 78 is out! I can confirm that the problem disappears in Chrome 78 :-)

**Comment by liweinan0423** [comment]
> > This is also affecting some of the drag and drop functionality in the various editor apps at HubSpot (that use react-dnd inside an iframe). Still digging into the extent of the problem.
> > And I have replicated the problem demonstrated by the [codesandbox example](https://codesandbox.io/s/github/react-dnd/react-dnd/tree/gh-pages/examples_hooks_ts/01-dustbin/single-target?from-embed) in Chrome 77.0.3865.75 on Windows
> 
> > We are using react-dnd in an web application which is hosted inside another webapplication using a cross-origin iframe.
> > We have noticed that since the last Chrome upgrade all the drag-drop functionality stopped working. Our app is working fine in older Chrome versions and in other browsers.
> > We are not sure whether this is hitting an edge case in this library or the issue is with in Chrome 77.0
> > **Reproduction**
> > Go to http://react-dnd.github.io/react-dnd/examples/dustbin/single-target (or to any of the examples)
> > Steps to reproduce the behavior:
> > The "inline" example is working fine:
> > ![chrome-working](https://user-images.githubusercontent.com/251330/64860544-c6e0bb80-d62d-11e9-8f0d-fed1c1f7c397.gif)
> > However running the same code in the JavaScript/TypeScript tabs, so in a cross-origin iframe (e.g. in the embedded codesandbox) the drop event is not fired:
> > ![chrome-not-working](https://user-images.githubusercontent.com/251330/64860595-e841a780-d62d-11e9-9b14-1e421b4c40c1.gif)
> > [Running the example](https://codesandbox.io/s/github/react-dnd/react-dnd/tree/gh-pages/examples_hooks_ts/01-dustbin/single-target?from-embed) is also not working in the CodeSandBox editor which also uses an inframe.
> > ![chrome-not-working-codesandbox](https://user-images.githubusercontent.com/251330/64860858-bbda5b00-d62e-11e9-833b-664c93989a23.gif)
> > In both cases there is this error in the console:
> > ```
> > browser.js:38 Uncaught Invariant Violation: Cannot call hover while not dragging.
> >     at invariant (https://hmq56.csb.app/node_modules/invariant/browser.js:38:15)
> >     at checkInvariants (https://hmq56.csb.app/node_modules/dnd-core/dist/esm/actions/dragDrop/hover.js:45:29)
> >     at DragDropManagerImpl.hover (https://hmq56.csb.app/node_modules/dnd-core/dist/esm/actions/dragDrop/hover.js:25:5)
> >     at Object.eval [as hover] (https://hmq56.csb.app/node_modules/dnd-core/dist/esm/DragDropManagerImpl.js:132:38)
> >     at HTML5Backend.handleTopDrop (https://hmq56.csb.app/node_modules/react-dnd-html5-backend/dist/esm/HTML5Backend.js:418:21)
> > ```
> > 
> > 
> > **Expected behavior**
> > No exceptions in the console, drop event is fired
> > **Desktop (please complete the following information):**
> > 
> > * OS: Windows 10 (1903)
> > * Browser Chrome
> > * Version 77.0.3865.75 (Official Build) (64-bit)
> > 
> > **Additional context**
> > We did some investigation and we've observed that Chrome fires the dragEnd event prematurely in the case of the iframe which breaks the invariant checks.
> 
> 请问这个问题解决了吗？

你用中文提问是没人回答你的

**Comment by moyarich** [comment]
I recently encountered this error while working on my project. Upon investigation, I discovered that another component on the page was causing the drop event to fire unexpectedly.

The issue arose because I was using both "react-dropzone" and "react-dnd" on the same page. Whenever I attempted to drop a file onto a dropzone, the error occurred. It seemed that the drop event triggered by react-dropzone was propagating and being captured by "react-dnd".

To resolve this issue, I had to make a modification when calling the useDropzone hook from "react-dropzone". Specifically, I set noDragEventsBubbling to false:

```javascript

    useDropzone({
        noDragEventsBubbling: false,
        ...props,
    })
```
This adjustment ensured that the drag events were handled appropriately, resolving the error.
